### PR TITLE
[dy] Support serialize Polars dataframe

### PR DESCRIPTION
**Pull request by dy46**
# Description

Resolves https://github.com/mage-ai/mage-ai/issues/4102.

In order to serialize a polars dataframe, we have to write some metadata about the output variable so that we can determine whether or not to use polars or pandas when serializing. This PR adds a `type.json` file to the variable directory that currently just stores the type of the variable which will be used when reading the output data. If there is no file, we will fallback to the previous logic.

# How Has This Been Tested?

- [x] Tested locally with the example in the issue
Screenshot 2024-01-03 at 5 33 36 PM
Screenshot 2024-01-03 at 5 33 32 PM
- [x] Tested with S3 remote variable directory
- [x] Unit test

# Checklist
- [x] The PR is tagged with proper labels (bug, enhancement, feature, documentation)
- [x] I have performed a self-review of my own code
- [x] I have added unit tests that prove my fix is effective or that my feature works
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] If new documentation has been added, relative paths have been added to the appropriate section of `docs/mint.json`

cc: @wangxiaoyou1993

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -25,7 +25,7 @@ ldap3==2.9.1
 newrelic==8.8.0
 numpy>=1.21.0
 pandas>=1.3.0
-polars<=0.19.2
```
Comment (wangxiaoyou1993) [inline requirements.txt:28]: did this version cause any issues?

File: mage_ai/data_preparation/storage/local_storage.py [hunk 2]
```diff
@@ -93,9 +93,12 @@ async def write_json_file_async(self, file_path: str, data) -> None:
             )
             await file.write(fcontent)
 
-    def read_parquet(self, file_path: str, **kwargs) -> pd.DataFrame:
```
Comment (wangxiaoyou1993) [inline mage_ai/data_preparation/storage/local_storage.py:96]: Let's not remove `**kwargs`. It'll cause errors.
The S3Storage and GCSStorage also have the `**kwargs` param https://github.com/mage-ai/mage-ai/blob/master/mage_ai/data_preparation/storage/s3_storage.py#L96

File: mage_ai/data_preparation/models/variable.py [hunk 3]
```diff
@@ -273,6 +302,20 @@ async def write_data_async(self, data: Any) -> None:
         else:
             await self.__write_json_async(data)
 
+        self.write_metadata()
+
+    def write_metadata(self) -> None:
```
Comment (wangxiaoyou1993) [inline mage_ai/data_preparation/models/variable.py:307]: we use another [`metadata.json`](https://github.com/mage-ai/mage-ai/blob/master/mage_ai/data_preparation/models/variable.py#L607) to store dataframe stats. can we clarify the difference here?

File: requirements.txt (hunk 1 above)
Comment (dy46) [reply requirements.txt:28]: the `.cast` function doesn't exist in 0.17.x, and the polars documentation doesn't even have a page for under 0.18, so i just set it as a lower bound

File: mage_ai/data_preparation/models/variable.py (hunk 3 above)
Comment (dy46) [reply mage_ai/data_preparation/models/variable.py:307]: ok i wasn't sure if that was still being used. i can change the name of the new file

**Comment by dy46** [comment]
@wangxiaoyou1993 Thanks for catching those issues. I tested it more with S3 and found some issues, so I fixed those as well as addressed your comments. I also changed the name of the file to `type.json` to differentiate it.

File: mage_ai/data_preparation/models/block/__init__.py [hunk 4]
```diff
@@ -2074,6 +2075,24 @@ async def __get_outputs_async(
                 )
                 data_products.append(data)
                 continue
+            if type(data) is pl.DataFrame:
+                row_count, column_count = data.shape
```
Comment (wangxiaoyou1993) [inline mage_ai/data_preparation/models/block/__init__.py:2079]: the row_count and column_count are sample data's row and column count.
if we want to get accurate values, we need store them.

File: mage_ai/data_preparation/models/block/__init__.py [hunk 5]
```diff
@@ -2074,6 +2075,24 @@ async def __get_outputs_async(
                 )
                 data_products.append(data)
                 continue
+            if type(data) is pl.DataFrame:
```
Comment (wangxiaoyou1993) [inline mage_ai/data_preparation/models/block/__init__.py:2078]: you should use `elif`

File: requirements.txt (hunk 1 above)
Comment (wangxiaoyou1993) [reply requirements.txt:28]: how about changing `<=0.19.2` to `<0.19.2`?

**Comment by dy46** [comment]
@wangxiaoyou1993 addressed your comments

**Review by wangxiaoyou1993 — APPROVED** [review]

**Pull request merged**